### PR TITLE
Perf/cuda compact top p replay

### PR DESCRIPTION
**Pull request by inaniloquentee**
## Summary

- preserve strict VIME/vLLM top-p semantics by replaying the native finite nucleus
- add an SM90 kernel that scores compact replay IDs directly instead of materializing a full-vocabulary keep mask
- preserve rank-ordered reductions and handle TP shards with no replay candidates
- retain main's temperature-scaling compatibility contract from #423

## Performance

H100 TP4/CP2, Qwen3-8B, temperature 0.7, top_p 0.95, rollout batch 1 x 8, with --use-rollout-logprobs disabled:

- rollout: 58.52 s -> 48.49 s (-17.1%)
- rollout throughput: 87.48 -> 106.02 tok/GPU/s (+21.2%)
- end-to-end step: 93.88 s -> 84.14 s (-10.4%)
- train/rollout mismatch count and max |delta logp| remain exactly zero

The three old/new G11 steps produced identical response lengths, so the rollout comparison is workload-matched.

## Validation

- pytest tests/test_vime_linear_logp_provider.py -q: 10 passed
- pytest tests/test_framework_runtime_adapters.py::test_vllm_logp_replaces_every_duplicate_sampled_token_column -q: passed
- randomized dense-mask vs compact-support CUDA checks: bitwise equal across TP offsets, including empty local shards
- 3-step Qwen3-8B strict R/R run: zero mismatches and zero max |delta logp| on every step
- CUDA 12.8 / SM90 build and 200-step validation run are in progress

## Summary by CodeRabbit

- **New Features**
  - Added deterministic scoring for compact top-p replay data, including temperature support and tensor-parallel execution.
  - Added a public interface for calculating target logits and log-sum-exp values from sparse top-p results.
  - Strict VIME log-probability processing now supports replay masks when reusable local logits are available.

- **Bug Fixes**
  - Empty vocabulary shards now produce stable zero, negative-infinity, and log-probability results instead of NaNs.

### CODE REVIEW

**Review by Flink-ddd — APPROVED** [review]
LGTM

**Review by maxiaosong1124 — APPROVED** [review]
LGTM!
